Sources/SwiftOperatorPrecedence/OperatorPrecedence+Folding.swift

```diff
@@ -0,0 +1,364 @@
+//===------------------ OperatorPrecedence+Folding.swift ------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+import SwiftSyntax
+
+extension OperatorPrecedence {
+  private struct PrecedenceBound {
+    let groupName: PrecedenceGroupName?
+    let isStrict: Bool
+    let syntax: SyntaxProtocol?
+  }
+
+  /// Determine whether we should consider an operator in the given group
+  /// based on the specified bound.
+  private func shouldConsiderOperator(
+    fromGroup groupName: PrecedenceGroupName?,
+    in bound: PrecedenceBound,
+    fromGroupSyntax: SyntaxProtocol?,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> Bool {
+    guard let boundGroupName = bound.groupName else {
+      return true
+    }
+
+    guard let groupName = groupName else {
+      return false
+    }
+
+    if groupName == boundGroupName {
+      return !bound.isStrict
+    }
+
+    return try precedence(
+      relating: groupName, to: boundGroupName,
+      startSyntax: fromGroupSyntax, endSyntax: bound.syntax,
+      errorHandler: errorHandler
+    ) != .lowerThan
+  }
+
+  /// Look up the precedence group for the given expression syntax.
+  private func lookupPrecedence(
+    of expr: ExprSyntax,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> PrecedenceGroupName? {
+    // A binary operator.
+    if let binaryExpr = expr.as(BinaryOperatorExprSyntax.self) {
+      let operatorName = binaryExpr.operatorToken.text
+      return try lookupOperatorPrecedenceGroupName(
+        operatorName, referencedFrom: binaryExpr.operatorToken,
+        errorHandler: errorHandler
+      )
+    }
+
+    // FIXME: Handle all of the language-defined precedence relationships.
+    return nil
+  }
+
+  /// Form a binary operation expression for, e.g., a + b.
+  private func makeBinaryOperationExpr(
+    lhs: ExprSyntax, op: ExprSyntax, rhs: ExprSyntax
+  ) -> ExprSyntax {
+    ExprSyntax(
+      InfixOperatorExprSyntax(
+        leftOperand: lhs, operatorOperand: op, rightOperand: rhs)
+      )
+  }
+
+  /// Determine the associativity between two precedence groups.
+  private func associativity(
+    firstGroup: PrecedenceGroupName?,
+    firstGroupSyntax: SyntaxProtocol?,
+    secondGroup: PrecedenceGroupName?,
+    secondGroupSyntax: SyntaxProtocol?,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> Associativity {
+    guard let firstGroup = firstGroup, let secondGroup = secondGroup else {
```

Comment:
Why not make `firstGroup` and `secondGroup` non-optional if the first thing we do is return if they are `nil`?

Reply:
That would force callers to effectively do the same computation, so it's better to have it done once here.

Reply:
I think I would have probably done it on the call side, but I’m fine to keep it as-is.